### PR TITLE
Add support for user-defined counters

### LINKED ISSUE #11: Add support for custom metrics
Some benchmarks require tracking domain-specific settings such as:

* Number of items processed
* Amount of data processed and/or processing throughput in mb/s
* Amount of examples in machine learning models
* Domain-specific custom performance metrics (e.g., number of JIT recompilations in tracing ML compilers)

To expose this we can provide an extra overload where benchmark obtains a handle to `BenchmarkState`:

```swift
benchmark("my benchmark") { state in 
   let result = doSomethingExpensive()
   state.metric(name: "examples", value: result.examples.count)
}
```

The framework should report be able to report both the raw metric and metrics divided by the running time (i.e., examples -> examples/s, items -> items/s).

### PR DESCRIPTION
**Pull request by shabalind**
This change introduces support for counters that be updated through either:

1) `state.counters["name"] = value`
2) `state.increment(counter: "name")`
3) `state.increment(counter: "name", by: value)`

Counters are useful to record domain-specific statistics apart from the time measurements.

Fixes #11

### CODE REVIEW

File: Sources/Benchmark/BenchmarkReporter.swift
Comment (saeta) [inline Sources/Benchmark/BenchmarkReporter.swift:29]: Being able to capture counters is good, but what about arbitrary custom statistics of the counters and/or timing metrics? (This is okay to defer to a subsequent PR; just want us to make sure we're not painting ourselves into a corner.) Examples include: `min`, `max`, and arbitrary percentiles, as they can each provide useful information.

File: Sources/Benchmark/BenchmarkState.swift
Comment (saeta) [inline Sources/Benchmark/BenchmarkState.swift]: What's the advantage of `Int` vs `Double`? 

Additionally, it might be a good idea to (perhaps in a subsequent PR) use some `Double`-like value (or some more sophisticated collection type) that aggregates across threads safely.

**Review by saeta — APPROVED** [review]

**Comment by saeta** [comment]
Approving eagerly to unblock progress.

File: Sources/Benchmark/BenchmarkReporter.swift
Comment (shabalind) [reply Sources/Benchmark/BenchmarkReporter.swift:29]: Semantics of counters is based on https://github.com/google/benchmark#custom-counters. Integer totals are nice for things like examples/s, bytes/s, items/s results where you are interested in providing one aggregate total count for the whole benchmark run. 

`min`, `max`, `percentileN`, `avg` is coming in a follow-up PR soon! They are not metrics, but rather aggregates of the recorded results.

Comment (saeta) [reply Sources/Benchmark/BenchmarkReporter.swift:29]: Right, but Google Benchmark counters are `double`-like in C++, not `Int`s.

File: Sources/Benchmark/BenchmarkState.swift
Comment (shabalind) [reply Sources/Benchmark/BenchmarkState.swift]: I believe counters solve all of our short-term needs for `swift-models` and they have precendent coming from google/benchmark (so it's nice to support them anyway). 

For `Double`-based measurements I would assume we record them as `[Double]`, and only aggregate in the reporting layer, similarly to time and warmup measurements (so there is no issues with aggregation across multiple threads). 

I think two APIs can co-exist, once we have more concrete use cases for `[Double]` we can add it alongside counters.

Comment (saeta) [reply Sources/Benchmark/BenchmarkState.swift]: Okay, works for me.

File: Sources/Benchmark/BenchmarkReporter.swift
Comment (shabalind) [reply Sources/Benchmark/BenchmarkReporter.swift:29]: I guess it makes sense to go for `Double` here as well then.

Comment (shabalind) [reply Sources/Benchmark/BenchmarkReporter.swift:29]: Counter type has been changed to `Double` in 36dd339.

**Pull request merged**
